Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityDevice.cs

```diff
@@ -0,0 +1,466 @@
+        {
+            //InteractionSourceUpdated(args.state);
+        }
+
+        private void SpatialManager_SourcePressed(Windows.UI.Input.Spatial.SpatialInteractionManager sender, Windows.UI.Input.Spatial.SpatialInteractionSourceEventArgs args)
+        {
+            //InteractionSourcePressed(args.state, args.pressType);
+        }
+        private void SpatialManager_SourceReleased(Windows.UI.Input.Spatial.SpatialInteractionManager sender, Windows.UI.Input.Spatial.SpatialInteractionSourceEventArgs args)
+        {
+            //InteractionSourceReleased(args.state, args.pressType);
+        }
+
+        private void SpatialManager_SourceLost(Windows.UI.Input.Spatial.SpatialInteractionManager sender, Windows.UI.Input.Spatial.SpatialInteractionSourceEventArgs args)
+        {
+            //RemoveWindowsMixedRealityController(args.state);
+        }
+
+        private void SpatialManager_InteractionDetected(Windows.UI.Input.Spatial.SpatialInteractionManager sender, Windows.UI.Input.Spatial.SpatialInteractionDetectedEventArgs args)
+        {
+            //Not Implemented Yet
+        }
+
+#endif // WINDOWS_UWP
+
+        #endregion Experimental_PLAYER native device input
+
+        #region Mixed Reality controller handlers
+
+        // TODO (understatement)
+        // 1 - Test
+        // 2 - Refactor for multi-controller types (possibly as part of the inputsource interface)
+
+        /// <summary>
+        /// Retrieve the source controller from the Active Store, or create a new device and register it
+        /// </summary>
+        /// <param name="interactionSourceState">Source State provided by the SDK</param>
+        /// <returns>New or Existing Controller Input Source</returns>
+        private IMixedRealityController GetOrAddWindowsMixedRealityController(InteractionSourceState interactionSourceState)
+        {
+            //If a device is already registered with the ID provided, just return it.
+            if (activeControllers.ContainsKey(interactionSourceState.source.id))
+            {
+                return activeControllers[interactionSourceState.source.id];
+            }
+
+            var controllingHand = interactionSourceState.source.handedness == InteractionSourceHandedness.Left ? Handedness.Left : Handedness.Right;
+
+            //TODO - Controller Type Detection?
+            //Define new Controller
+            var detectedController = new WindowsMixedRealityController(
+                ControllerState.NotTracked,
+                controllingHand,
+                inputSystem.RequestNewGenericInputSource($"Mixed Reality Controller {controllingHand}"),
+                new Dictionary<DeviceInputType, InteractionMapping>()
+                );
+
+            detectedController.SetupInputSource(interactionSourceState);
+
+            activeControllers.Add(interactionSourceState.source.id, detectedController);
```

Comment:
This is the boxing I'm taking about. Not a big deal.

Reply:
I'd like to avoid boxing if at all possible... It can generate garbage VERY quickly.

Reply:
I think it's not that big a deal here since we only create a few devices at a time. Generally this won't happen that often.